tokio/src/sync/mpsc/block.rs

```diff
@@ -319,6 +347,7 @@ impl<T> Block<T> {
             None => {
                 // The compare-and-swap succeeded and the newly allocated block
                 // is successfully pushed.
+                self.set_has_next();
                 return new_block;
             }
```

Comment:
Another thread could see the compare and swap before this `set_has_next` has completed. Is that a problem?

Reply:
Trying to page all of that back as it has been a while. I had a thought about that when I was writing it and I _think_ it is ok. I think what happens in the situation that you are describing is that for example if you have two blocks such:
       
```
        B1                               B2
| a | b | c | _ | _ | -------> | _ | _ | _ | _ |
```

If we are trying to read at the second to last slot of B1 and a value is being written at the first slot of B2 and the CAS has succeeded and the thread reading B1 observes that, this means that we have a block but a value will not be written to this block before the current one is marked as having a next block. In this case, we will receive an Empty error, which will technically be correct because the value is still not there. Does that sound correct or I am making things up?

Reply:
If the consequence of this is that it reports `Empty` if a send is happening concurrently, then that's fine.

Reply:
Yes, that is correct.

Reply:
@Darksonn is that PR still relevant?  Is there plan to merge it at some point. If not we can close it I guess.

Reply:
We do want to add back `try_recv`, but I'm not sufficiently familiar with the internals of the mpsc channel to properly review it.

Reply:
Got it, makes sense. Thanks for looking at it nevertheless